File: apps/mobile/ios/BeaconExport/ExportFileStore.swift
import Foundation

/// Owns the temporary directory the export module writes into, and is the only thing that
/// decides whether a path belongs to it (PROJECT.md 21).
///
/// Pure Swift with no UIKit dependency: writing a document and deciding whether a path is
/// ours is plain file work, so it is unit tested. Presenting the result is a separate job and
/// lives in `ShareSheetPresenter`.
///
/// The store keeps a named subdirectory of the app's temporary directory rather than using
/// that directory directly, so `clear()` can never remove something another part of the app
/// left there.
final class ExportFileStore {

  /// Absolute URL of the directory this store owns. It is created on the first write; until
  /// then it simply does not exist, which is not an error anywhere in this file.
  let directory: URL

  private let fileManager: FileManager

  init(directory: URL = ExportFileStore.defaultDirectory, fileManager: FileManager = .default) {
    self.directory = directory
    self.fileManager = fileManager
  }

  /// The directory the app uses: `<tmp>/beacon-exports`.
  static var defaultDirectory: URL {
    URL(fileURLWithPath: NSTemporaryDirectory(), isDirectory: true)
      .appendingPathComponent("beacon-exports", isDirectory: true)
  }

  /// Writes `contents` as UTF-8 and returns the file's URL, replacing any file already there
  /// under the same name.
  ///
  /// Only the last path component of `fileName` is used: the caller names a file, it does not
  /// get to choose where in the container that file lands.
  func write(fileName: String, contents: String) throws -> URL {
    let name = (fileName as NSString).lastPathComponent
    // `lastPathComponent` reduces "../" to ".." and "/" to "/", both of which would still
    // point outside the directory, so those names are refused rather than sanitised further.
    guard !name.isEmpty, name != ".", name != "..", !name.contains("/") else {
      throw ExportError(code: .writeFailed, message: "\"\(fileName)\" is not a usable file name")
    }
    do {
      try fileManager.createDirectory(at: directory, withIntermediateDirectories: true, attributes: nil)
      let url = directory.appendingPathComponent(name, isDirectory: false)
      try contents.write(to: url, atomically: true, encoding: .utf8)
      return url
    } catch {
      throw ExportError(code: .writeFailed, message: (error as NSError).localizedDescription)
    }
  }

  /// True when `path` names something inside this store's directory.
  ///
  /// Both sides are standardised and symlink-resolved before they are compared, so a `..` in
  /// the middle of the path cannot walk out of the directory and back in through a name that
  /// looks right. The comparison is component by component rather than on the string, so a
  /// sibling directory whose name merely starts with ours does not pass either.
  func contains(path: String) -> Bool {
    let root = ExportFileStore.normalized(directory).pathComponents
    let candidate = ExportFileStore.normalized(URL(fileURLWithPath: path)).pathComponents
    guard candidate.count > root.count else { return false }
    return Array(candidate.prefix(root.count)) == root
  }

  /// True when `path` names a file that is actually on disk. Kept apart from
  /// `contains(path:)` because the two answer different questions, even though the caller
  /// reports the same error for both: a path we never wrote and a path that has since been
  /// cleared are equally unshareable.
  func exists(path: String) -> Bool {
    fileManager.fileExists(atPath: path)
  }

  /// Removes everything in the directory and returns how many entries were removed.
  ///
  /// Never fails. A directory that was never created, or a file that disappeared underneath
  /// us, both leave the caller with what it asked for, so there is nothing to report.
  func clear() -> Int {
    let contents = (try? fileManager.contentsOfDirectory(
      at: directory,
      includingPropertiesForKeys: nil,
      options: []
    )) ?? []
    var removed = 0
    for url in contents {
      do {
        try fileManager.removeItem(at: url)
        removed += 1
      } catch {
        // Something else got there first. The file is gone either way; it just was not us.
        continue
      }
    }
    return removed
  }

  /// `standardized` is the documented removal of `.` and `..`, `resolvingSymlinksInPath` the
  /// documented resolution of links: a traversal has to survive both to be counted as ours.
  ///
  /// A path that is not on disk can normalise differently from one that is, because symlink
  /// resolution leaves components it cannot find alone. That only ever loses a match, and a
  /// file that is not there is refused anyway, so it costs nothing here.
  private static func normalized(_ url: URL) -> URL {
    url.standardized.resolvingSymlinksInPath()
  }
}

File: apps/mobile/android/app/src/main/java/com/beacon/export/ExportFileStore.kt
package com.beacon.export

import java.io.File

/**
 * Owns the one directory the export module is allowed to touch, and is the only place that
 * decides whether a path belongs to it.
 *
 * Free of Android framework types so the containment rules are unit tested on the JVM; the
 * module supplies the cache directory and turns the results into bridge values.
 */
class ExportFileStore(cacheDir: File, directoryName: String = DIRECTORY_NAME) {

    /** The subdirectory this store owns. Created lazily, on the first write. */
    val directory: File = File(cacheDir, directoryName)

    /**
     * Writes [contents] as UTF-8 and returns the file.
     *
     * [fileName] comes from JavaScript, so only its last path component is used: a caller that
     * passes `../../databases/beacon.db` gets `beacon.db` inside this directory rather than a
     * write outside it. An existing file of the same name is replaced, because the export is a
     * snapshot of the session as it stands and keeping the previous one would only leave the
     * user two files to choose between.
     */
    fun write(fileName: String, contents: String): File {
        val safeName = File(fileName).name
        // `.` and `..` survive the reduction above and name a directory, not a file; refusing
        // them here keeps the failure a clear argument error rather than an IO error later.
        require(safeName.isNotBlank() && safeName != "." && safeName != "..") {
            "A file name is required"
        }
        directory.mkdirs()
        val target = File(directory, safeName)
        target.writeText(contents, Charsets.UTF_8)
        return target
    }

    /**
     * Whether [path] names a file this store owns, i.e. a direct child of [directory].
     *
     * Both sides are canonicalised first so a path that walks out and back in (`beacon-exports/
     * ../../databases/beacon.db`) is compared as where it actually lands. This is what stops the
     * module granting a share target read access to an arbitrary file: JavaScript hands over a
     * path, and nothing but this check limits which one.
     *
     * Existence is deliberately not part of the answer; the caller checks that separately so a
     * missing file and a path outside the directory can be told apart while debugging.
     */
    fun contains(path: String): Boolean {
        if (path.isBlank()) return false
        return runCatching {
            File(path).canonicalFile.parentFile == directory.canonicalFile
        }.getOrDefault(false)
    }

    /**
     * Deletes everything in [directory] and returns how many entries went away.
     *
     * A directory that was never created is simply empty, so clearing twice — or before the
     * first export — is a no-op that reports zero rather than a failure.
     */
    fun clear(): Int {
        val entries = directory.listFiles() ?: return 0
        // Only plain files are ever written here, but deleting recursively means a stray
        // directory cannot wedge the clear and leave the count wrong on every later call.
        return entries.count { it.deleteRecursively() }
    }

    companion object {
        /** Named after the app so it is recognisable when browsing the cache during support. */
        const val DIRECTORY_NAME = "beacon-exports"
    }
}
